go-controller/pkg/clustermanager/healthcheck/provider.go

```diff
@@ -0,0 +1,155 @@
+package healthcheck
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+
+	coreinformers "k8s.io/client-go/informers/core/v1"
+)
+
+type HealthState int
+
+const (
+	// UNKNOWN is the health state of any node that is not being monitored.
+	// Services should not make any determination based on this state and should
+	// consider the node potentially available for scheduling.
+	UNKNOWN HealthState = iota
+
+	// AVAILABLE is the health state of a node that is reachable and thought to
+	// have a healthy ovn-kube running. Services should consider the node
+	// available for scheduling.
+	AVAILABLE
+
+	// UNREACHABLE is the health state of a node that is not reachable. Services
+	// should not consider the node available for scheduling and should attempt
+	// to unschedule from it as well.
+	UNREACHABLE
+)
+
+func (s HealthState) String() string {
+	switch s {
+	case AVAILABLE:
+		return "AVAILABLE"
+	case UNREACHABLE:
+		return "UNREACHABLE"
+	}
+	return "UNKNOWN"
+}
+
+// Provider of node health state information
+type Provider interface {
+	// Register a consumer to be informed about node health state changes
+	Register(consumer Consumer)
+	// GetHealthState of the node
+	GetHealthState(node string) HealthState
+	// MonitorHealthState notifies interest of a consumer to monitor the health
+	// state of a node.
+	MonitorHealthState(node string)
+}
+
+// Consumer of node health state information
+type Consumer interface {
+	// Name the consumer goes by for logging purposes
+	Name() string
+	// HealthStateChanged informs consumers of node health state changes. A
+	// consumer might be informed about a node it is not interested in and
+	// should be able to ignore it in that case.
+	HealthStateChanged(node string)
+	// HealthStateChanged queries the interest of the consumer of being informed
+	// about health state changes of the node. A node is monitored if any
+	// consumer is interested. Monitoring for a node might terminate as soon as
+	// no consumer is interested.
+	MonitorHealthState(node string) bool
```

Comment:
so how does a consumer actually use this function? It's not clear to me what it is for.

Reply:
So, to use an simplified example, the egress ip controller uses `MonitorHealthState` to notify the healtcheck controller that the provided node should be probed because. let's say, the node was annotated with `egress-assignable`.

Reply:
So this is a function provided by the consumer to trigger ad-hoc healthchecks? If that's right, then can we change the comment and change the name of the function. Maybe TriggerProbe or something.

Reply:
Sorry, was wrong. The github review interface played a trick on me. Let's try again.

This function is implemented by consumer. It is used by the provider (the healcheck contoller) to ask the consumer if a node should be monitored. 

So this is how it works
1. A consumer asks the provider to monitor a node
    `consumer -> MonitorHealthState(node) -> provider`

2. As the provider monitors the node, it informs the consumer of health state changes
    `provider -> HealthStateChanged(node) -> consumer`
    It also asks the consumer if it is still interested in monitoring that node
   `provider -> MonitorHealthState(node) bool -> consumer`

3. A consumer, after being notified that the health state of a node change, can query the actual healthstate of the node
   `consumer -> GetHealthState(node) HealthState -> provider`

If it helps, I can add this flow in the provider doc.